Search for desktop files in all subdirectories of app dirs

yawor opened the pull request:
Fixes #367

yawor commented:
Do not merge yet. I need to fix something

yawor commented:
Ok, it's fixed now. I've moved the seen directory check to be better optimized.

pol-rivero commented on nwg_panel/icons.py:
```diff
@@ -70,10 +70,19 @@ def __process_desktop_file(file_path):
+                if os.path.isdir(file_path):
+                    app_dirs.insert(0, file_path)
+                    continue
                 __process_desktop_file(file_path)
```
nit: you could add
```
if not os.path.exists(file_path):
    continue
```
to avoid a crash on launch if the directory contains a broken symlink

yawor replied:
Very good idea. I haven't thought of that case.

yawor replied:
@pol-rivero I've also been thinking whether or not check if the file has a `.desktop` extension before passing it to `__process_desktop_file` method.

pol-rivero replied:
I also had the same doubt when writing it originally, by decided not to check it for 2 reasons:
1. It's not necessary. If our code works fine without this restriction, we shouldn't enforce it (see [Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle)).
2. The [Desktop Entry Specification](https://specifications.freedesktop.org/desktop-entry-spec/latest/file-naming.html) technically _does_ allow desktop files to use extensions other than `.desktop` (notice the use of "Desktop entry files **should** have the .desktop extension" instead of "**must**" or equivalent).

yawor commented:
I've added checking if the file exists.
I've also added handing of OSError when listing files in the directory in case the directory permissions don't allow listing its content by the current user. The directory is skipped in that case.
Added OSError handling in `__process_desktop_file` method when opening the file to be sure it can be opened.
Added UnicodeDecodeError in `__process_desktop_file` method when reading from opened file, as it can be raised if a binary file is opened in text mode (just to cover some edge cases that user maybe put some binary file there by accident).

pol-rivero commented:
Sounds good! This should now cover lots of edge cases. Looks good to merge.

nwg-piotr commented:
Very well then.
